appsec/src/extension/user_tracking.c

```diff
@@ -115,3 +154,104 @@ void dd_find_and_apply_verdict_for_user(zend_string *nonnull user_id)
         }
     }
 }
+
+bool dd_parse_user_collection_mode(
+    zai_str value, zval *nonnull decoded_value, bool persistent)
+{
+    if (!value.len) {
+        return false;
+    }
+
+    size_t len = strlen((const char *)value.ptr);
+    if (dd_string_equals_lc(value.ptr, len, ZEND_STRL("ident")) ||
+        dd_string_equals_lc(value.ptr, len, ZEND_STRL("identification")) ||
+        dd_string_equals_lc(value.ptr, len, ZEND_STRL("extended"))) {
+        _user_mode = user_mode_ident;
+    } else if (dd_string_equals_lc(value.ptr, len, ZEND_STRL("anon")) ||
+               dd_string_equals_lc(
+                   value.ptr, len, ZEND_STRL("anonymization")) ||
+               dd_string_equals_lc(value.ptr, len, ZEND_STRL("safe"))) {
+        _user_mode = user_mode_anon;
+    } else { // If the value is disabled or an unknown value, we disable user ID
+             // collection
+        _user_mode = user_mode_disabled;
+    }
+
+    ZVAL_STR(decoded_value, zend_string_init(value.ptr, value.len, persistent));
+
+    return true;
+}
+
+zend_string *nullable dd_user_id_anonymize(zend_string *nonnull user_id)
+{
+    zend_string *digest;
+    const php_hash_ops *ops;
+    void *context;
+
+#if PHP_VERSION_ID < 80000
+    if (!_hash_fetch_ops) {
+        return NULL;
+    }
+
+    ops = _hash_fetch_ops(
+        ZSTR_VAL(_sha256_algo_zstr), ZSTR_LEN(_sha256_algo_zstr));
+#else
+    ops = php_hash_fetch_ops(_sha256_algo_zstr);
+#endif
+    if (!ops) {
+        mlog(dd_log_debug, "Failed to load sha256 algorithm");
+        return NULL;
+    }
+
+#if PHP_VERSION_ID < 80000
+    context = emalloc(ops->context_size);
+#else
+    context = php_hash_alloc_context(ops);
+#endif
+
+#if PHP_VERSION_ID < 80100
+    ops->hash_init(context);
+#else
+    ops->hash_init(context, NULL);
+#endif
+
+    ops->hash_update(
+        context, (unsigned char *)ZSTR_VAL(user_id), ZSTR_LEN(user_id));
+
+    digest = zend_string_alloc(ops->digest_size, 0);
+    ops->hash_final((unsigned char *)ZSTR_VAL(digest), context);
```

Comment:
more of a comment on poor spec drafting, but this is very bad anonymization. You can just look up a rainbow table and figure out a significant portion of the usernames